core/pool_vector.h

```diff
@@ -297,103 +146,140 @@ class PoolVector {
 
 	class Read : public Access {
 	public:
-		_FORCE_INLINE_ const T &operator[](int p_index) const { return this->mem[p_index]; }
-		_FORCE_INLINE_ const T *ptr() const { return this->mem; }
-
-		void operator=(const Read &p_read) {
-			if (this->alloc == p_read.alloc) {
-				return;
-			}
-			this->_unref();
-			this->_ref(p_read.alloc);
+		// This will go down if _alloc is NULL, (e.g. after calling release).
+		// There is no protection (perhaps for speed?)
+		// This was the same in the old PoolVector.. we could check for this.
+		_FORCE_INLINE_ const T &operator[](int p_index) const { return this->_alloc->vector.ptr()[p_index]; }
+		_FORCE_INLINE_ const T *ptr() const {
+			return this->_alloc ? this->_alloc->vector.ptr() : nullptr;
 		}
 
```

Comment:
I saw you've omitted the `operator=` overload here. Will that work ok?

If you assign e.g. a Read to Read, could you end up with two `Access`es which attempt to `unlock()` when they destruct, but `lock()` has only been called once?

Reply:
Yes there's even a few more things missing potentially as well...    It was mostly a draft PR based on your PR just to illustrate a simple way to cleanup the thread safety model so it's clear whats critical and what interface is thread safe.

Key points were more or less the above, i was just going to keep as a draft for any help or reference.